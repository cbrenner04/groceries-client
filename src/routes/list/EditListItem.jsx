import React from 'react';
import Async from 'react-async';
import PropTypes from 'prop-types';

import { fetchItemToEdit } from './utils';
import EditListItemForm from './containers/EditListItemForm';
import Loading from '../../components/Loading';
import UnknownError from '../error_pages/UnknownError';

function EditListItem(props) {
  return (
    <Async
      promiseFn={fetchItemToEdit}
      itemId={props.match.params.id}
      listId={props.match.params.list_id}
      history={props.history}
    >
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data) => (
          <EditListItemForm
            history={props.history}
            listUsers={data.listUsers}
            userId={data.userId}
            list={data.list}
            item={data.item}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
}

EditListItem.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
      list_id: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
};

export default EditListItem;
