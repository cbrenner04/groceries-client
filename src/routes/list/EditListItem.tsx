import React from 'react';
import Async, { type PromiseFn } from 'react-async';
import { useNavigate, useParams } from 'react-router-dom';

import { fetchItemToEdit } from './utils';
import EditListItemForm from './containers/EditListItemForm';
import Loading from '../../components/Loading';
import UnknownError from '../error_pages/UnknownError';
import type { IList, IListItem, IListUser } from '../../typings';

interface IFetchResponse {
  list: IList;
  item: IListItem;
  listUsers: IListUser[];
  userId: string;
}

const EditListItem: React.FC = (): React.JSX.Element => {
  const navigate = useNavigate();
  const { id, list_id: listId } = useParams();

  return (
    <Async promiseFn={fetchItemToEdit as unknown as PromiseFn<void>} itemId={id} listId={listId} navigate={navigate}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data: IFetchResponse): React.JSX.Element => (
          <EditListItemForm
            navigate={navigate}
            listUsers={data.listUsers}
            userId={data.userId}
            list={data.list}
            item={data.item}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
};

export default EditListItem;
