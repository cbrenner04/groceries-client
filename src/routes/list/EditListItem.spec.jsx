import React from 'react';
import { act, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import EditListItem from './EditListItem';
import axios from '../../utils/api';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
  useParams: () => ({
    id: '1',
    list_id: '1',
  }),
}));

describe('EditListItem', () => {
  const renderEditListItem = (newProps) => {
    return render(
      <MemoryRouter>
        <EditListItem {...newProps} />
      </MemoryRouter>,
    );
  };

  it('renders the Loading component when fetch request is pending', async () => {
    const { container, findByText } = renderEditListItem();
    const status = await findByText('Loading...');

    expect(container).toMatchSnapshot();
    expect(status).toBeTruthy();
  });

  it('displays UnknownError when an error occurs', async () => {
    axios.get = jest.fn().mockRejectedValue({ message: 'failed to send request' });
    const { container, getByRole } = renderEditListItem();

    await act(async () => undefined);

    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
    expect(getByRole('button')).toHaveTextContent('refresh the page');
  });

  it('displays EditListItem', async () => {
    axios.get = jest.fn().mockResolvedValue({
      data: {
        item: {
          user_id: 'id1',
          id: 'id1',
          product: 'foo',
        },
        list: {
          id: 'id1',
          type: 'GroceryList',
          name: 'foo',
          completed: false,
          refreshed: false,
          created_at: 'some date',
          owner_id: 'id1',
        },
        list_users: [{ id: 'id1', email: 'foo@example.com' }],
        categories: [],
      },
    });
    const { container, getByText } = renderEditListItem();

    await act(async () => undefined);

    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
    expect(getByText('Update Item')).toBeVisible();
  });
});
