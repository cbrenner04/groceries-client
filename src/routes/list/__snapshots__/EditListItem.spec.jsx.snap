// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditListItem displays EditListItem 1`] = `
<div>
  <h1>
    Edit 
    foo
  </h1>
  <br />
  <form
    autocomplete="off"
    class=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="product"
      >
        Product
      </label>
      <input
        class="form-control"
        id="product"
        name="product"
        placeholder="apples"
        type="text"
        value="foo"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="quantity"
      >
        Quantity
      </label>
      <input
        class="form-control"
        id="quantity"
        name="quantity"
        placeholder="3 bags"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="category"
      >
        Category
      </label>
      <input
        class="form-control"
        id="category"
        list="categories"
        name="category"
        type="text"
        value=""
      />
      <datalist
        id="categories"
      />
    </div>
    <div
      class="mb-3 form-group"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="purchased"
          name="purchased"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="purchased"
          title=""
        >
          Purchased
        </label>
      </div>
    </div>
    <button
      class="btn btn-success btn-block"
      type="submit"
    >
      Update Item
    </button>
    <a
      class="btn btn-link btn-block"
      href="/lists/1"
    >
      Cancel
    </a>
  </form>
</div>
`;

exports[`EditListItem displays UnknownError when an error occurs 1`] = `
<div>
  <h1>
    Something went wrong!
  </h1>
  <h2>
    We are currently unable to render this page.
  </h2>
  <p>
    Please check your connection and try to Â 
    <button
      style="background-color: transparent; cursor: pointer; display: inline; margin: 0px; padding: 0px; color: rgb(0, 123, 255); text-decoration: none;"
    >
      refresh the page
    </button>
  </p>
</div>
`;

exports[`EditListItem renders the Loading component when fetch request is pending 1`] = `
<div>
  <div
    style="display: flex; justify-content: center; align-items: center;"
  >
    <div
      class="spinner-border"
      role="status"
      style="width: 100px; height: 100px; position: absolute; top: 50%;"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
